crypto/sm2/sm2_za.c

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright 2017 The OpenSSL Project Authors. All Rights Reserved.
+ * Copyright 2017 Ribose Inc. All Rights Reserved.
+ * Ported from Ribose contributions from Botan.
+ *
+ * Licensed under the OpenSSL license (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/sm2.h>
+#include <openssl/evp.h>
+#include <string.h>
+
+int SM2_compute_za(uint8_t *out,
+                   const EVP_MD *digest, const char *user_id, const EC_KEY *key)
+{
+    int rc = 0;
+
+    const EC_GROUP *group = EC_KEY_get0_group(key);
+
+    BN_CTX *ctx = NULL;         // ???
```

Comment:
This seems pointless at the moment, you're just using it to pass a NULL parameter to EC_GROUP_get_curve_GFp() and EC_POINT_get_affine_coordinates_GFp(). You might as well either pass NULL directly to the functions without a variable if you want them to handle creating the BN_CTX or you should call BN_CTX_new() and BN_CTX_free() within this function and pass a valid pointer to the functions. The later will be more efficient than the current implementation as you'll only be doing a single BN_CTX_new()/BN_CTX_free() rather than 3.